### PR TITLE
Added ControlMode feature to ManualCameraControl (fly or walk) + made GestureManager editor friendly

### PR DESCRIPTION
**Pull request by stbertou**
- fly: current behaviour where the camera just moves along the forward direction
- walk: the camera will move on a X/Z plane, you can then also move up and down using page up/down
Also exposed 2 control speed values that used to be hardcoded
All default values have been set to not change the current behaviour of the script

Also trivial change to GestureManager: added editor only code to send the "OnSelect" message to the focused object when we right click or press a specific key (space being the default)

### CODE REVIEW

File: Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs [hunk 1]
```diff
@@ -28,6 +28,15 @@ public enum MouseButton
     public MouseButton MouseLookButton = MouseButton.Shift;
     public bool IsControllerLookInverted = true;
 
+    public enum ControlMode
+    {
+        Fly,
+        Walk,
+    }
+    public ControlMode controlMode = ControlMode.Fly;
+    public float controlSlowSpeed = 0.1f;
```
Comment (jwittner) [inline Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs]: These should be PascalCase if they're to be public, e.g. ControlSlowSpeed, ControlMode.

File: Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs [hunk 2]
```diff
@@ -89,7 +106,7 @@ private Vector3 GetCameraControlTranslation()
         deltaPosition += InputCurve(Input.GetAxis("Horizontal")) * this.transform.right;
         deltaPosition += InputCurve(Input.GetAxis("Vertical")) * this.transform.forward;
 
-        float accel = Input.GetKey(KeyCode.LeftShift) ? 1.0f : 0.1f;
+        float accel = Input.GetKey(KeyCode.LeftShift) ? controlFastSpeed : controlSlowSpeed;
```
Comment (jwittner) [inline Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs]: Nice! =)

File: Assets/HoloToolkit/Input/Scripts/GestureManager.cs [hunk 3]
```diff
@@ -47,14 +52,19 @@ void Start()
             gestureRecognizer.StartCapturingGestures();
         }
 
-        private void GestureRecognizer_TappedEvent(InteractionSourceKind source, int tapCount, Ray headRay)
+        private void SendMessageToFocusedObject()
```
Comment (jwittner) [inline Assets/HoloToolkit/Input/Scripts/GestureManager.cs]: Maybe something like 'OnTap' would be a more clear name here? My real qualm is that this doesn't just send any message to the focused object it sends 'OnSelect' to the focused object. I think 'OnTap' makes it clear that the method should be called whenever the GestureManager recognizes a tap.

File: Assets/HoloToolkit/Input/Scripts/GestureManager.cs [hunk 4]
```diff
@@ -15,6 +15,11 @@ namespace HoloToolkit.Unity
     public partial class GestureManager : Singleton<GestureManager>
     {
         /// <summary>
+        /// Key to press in the editor to select the currently gazed hologram
+        /// </summary>
+        public KeyCode editorSelectKey = KeyCode.Space;
```
Comment (jwittner) [inline Assets/HoloToolkit/Input/Scripts/GestureManager.cs]: PascalCase here as well.

File: Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs [hunk 5]
```diff
@@ -28,6 +28,18 @@ public enum MouseButton
     public MouseButton MouseLookButton = MouseButton.Shift;
     public bool IsControllerLookInverted = true;
 
+    public enum ControlMode
```
Comment (stbertou) [inline Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs]: ManualCameraControl should be moved to the HoloToolKit namespace so that we can move this enum outside the class scope and just rename the member back to just be ControlMode
Happy to do that now or it could be done in a seperate step

Comment (jwittner) [inline Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs]: It does look like there's quite the mix of being included in a namespace and not. My vote is that we should unless there's a compelling reason not to. Maybe hold off on it for  now and make an issue pointing out this inconsistency. That way if there's some reason for it there's a place for that to come up.

**Comment by jwittner** [comment]
Pulling down latest and giving it a test.

File: Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs [hunk 6]
```diff
@@ -89,7 +109,7 @@ private Vector3 GetCameraControlTranslation()
         deltaPosition += InputCurve(Input.GetAxis("Horizontal")) * this.transform.right;
         deltaPosition += InputCurve(Input.GetAxis("Vertical")) * this.transform.forward;
```
Comment (jwittner) [inline Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs]: I think you need the same control mode checks for the "Vertical" axis. Testing movement in the ManualCameraControl scene, the arrow keys work correctly, but WASD controls behave as they used to. I believe they should both respect the control mode.

I also noticed that Shift is the hard coded key binding for fast/slow, which makes sense because it matches the unity editor, but we've recently changed the default mouse control to Shift as well. We probably need to make the fast/slow controls configurable or take away Shift for mouse control even though it's the best option given the unsaved scene + waSd dilemma. I'm not sure, but what we have now is confusing and by defaulting to Shift movement with mouse control is always fast. Maybe I should make an issue for this outside of your change?

**Comment by jwittner** [comment]
I noticed that Shift is the hard coded key binding for fast/slow, which makes sense because it matches the unity editor, but we've recently changed the default mouse control to Shift as well. We probably need to make the fast/slow controls configurable or take away Shift for mouse control even though it's the best option given the unsaved scene + waSd dilemma. I'm not sure, but what we have now is confusing and by defaulting to Shift movement with mouse control is always fast. Maybe I should make an issue for this outside of your change?

**Comment by jwittner** [comment]
Tested the Gesture Manager changes and they seem to work well!

**Comment by stbertou** [comment]
great! what's the next step?

File: Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs (hunk 5 above)
Comment (stbertou) [inline Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs]: should I log a new issue about that ? then it can be tacked for everything not just ManualCameraControl ?

**Comment by stbertou** [comment]
Happy to expose this hard coded key and default to something else that SHIFT
Maybe right control ?

File: Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs (hunk 6 above)
Comment (stbertou) [inline Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs]: I agree. I'll fix/clean all that no worries.
I'll refactor the code as well as we either use the down/up keys or the horizontal/vertical axis so we should multiply by the transform vectors just once

File: Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs (hunk 5 above)
Comment (NeerajW) [inline Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs]: Trying to understand the issue: Is it that the class is not in the HoloToolkit.Unity namespace?

Comment (jwittner) [inline Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs]: I'm thinking that ControlMode makes more sense inside the class, but also that yes, we should move the class inside the standard namespace.

Comment (stbertou) [inline Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs]: @NeerajW yes indeed this class (ManualCameraControl) is not in the HolotoolKit namespace
Happy to "fix that" but it will probably make more sense to do this change in another PR
@jwittner sounds good then even if we keep it inside the class like that we'll have to keep "CurrentControlMode" which I guess is not a big deal anyway.

**Comment by jwittner** [comment]
I'm thinking maybe a mouse button might be best for enabling mouse control of the camera. Personally I'd preferred a 1:1 mapping to what the emulator has, but I don't recall off hand how that works - I'll have to check it out tomorrow. =)

File: Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs (hunk 5 above)
Comment (NeerajW) [inline Assets/HoloToolkit/Utilities/Scripts/ManualCameraControl.cs]: I like this. Having both enum and class inside the HoloToolkit.Unity for better discoverability.

**Comment by NeerajW** [comment]
>>Personally I'd preferred a 1:1 mapping to what the emulator has

Totally agree so developers don't need a new mental map every time :)

**Comment by jwittner** [comment]
Hey @stbertou if you're going to bite off the keybinding issues hereI'm thinking of merging in #95 because it works well and is relatively small. That'll likely mean you'll have to merge in master again. Given that this has been open longer and touches the same code I wanted to give you a heads up.

**Comment by NeerajW** [comment]
@stbertou yeah please merge #95 and then let's merge your PR. :)

**Comment by stbertou** [comment]
FYI I've just rebased master so I should have #95 changes

**Comment by Subere23** [comment]
That confusion is my fault I think.I am still trying to figure out this
system in this type of group. I thought I reverted any of the commits I had
on your branch. Sorry for that.

On Thu, Jul 7, 2016 at 11:19 AM, Stephane Bertout <notifications@github.com>
wrote:

> FYI I've just rebased master so I should have #95
> <https://github.com/Microsoft/HoloToolkit-Unity/pull/95> changes
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/HoloToolkit-Unity/pull/83#issuecomment-231130208>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AG3e5YPpPaj0DMgclzhkg2OLY-yormc-ks5qTSb1gaJpZM4JA-4P>
> .
>



-- 
Jason Odom
*Thought Experiment*
205-223-1011
Jasonodom23@gmail.com

File: Assets/HoloToolkit/Build/Editor/BuildDeployPortal.cs [hunk 7]
```diff
@@ -0,0 +1,438 @@
+﻿//
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (NeerajW) [inline Assets/HoloToolkit/Build/Editor/BuildDeployPortal.cs]: This happened because you updated your fork with Master? :)

The pull request should have only updated itself in the event that you had to check in those changes to your branch. Which should not have happened.

I suspect we might have to close this pull and open a new one with only your updates.

**Comment by stbertou** [comment]
Yeah no worries I'm like you, never usd github this way before so I'm also learning ;-)
Anyway this PR has just been rebased with master so it should everything we want
Weirdly though the Files Changed tab doesn't contain just what I've changed, not sure why..

File: Assets/HoloToolkit/Build/Editor/BuildDeployPortal.cs (hunk 7 above)
Comment (stbertou) [inline Assets/HoloToolkit/Build/Editor/BuildDeployPortal.cs]: That wouldn't be an issue at all my changes were very basic anyway.
Up to you !

Comment (stbertou) [inline Assets/HoloToolkit/Build/Editor/BuildDeployPortal.cs]: But for the next time what is the process when master has been updated while you've been working on a specific PR ?

Comment (NeerajW) [inline Assets/HoloToolkit/Build/Editor/BuildDeployPortal.cs]: I will merge and see if there are any conflicts. I just am not sure if your commit should have shown all the other files everyone else changed.

**Comment by NeerajW** [comment]
Looks like merge worked. All set!

**Comment by stbertou** [comment]
Perfect ! thanks @NeerajW for that !

**Comment by jwittner** [comment]
Not sure what happened, but I agree it's really odd that the contents of master showed up in this PR if they really didn't change. @stbertou did you have any issues with the merge before the commit?

**Comment by stbertou** [comment]
No the merge went well.
Anyway everything seems fine now.
Thanks all !
